### ISSUE TITLE
Support actions of type Script or support pre and post steps in composite actions

### ISSUE THREAD

**Issue by umarcor** [actions/runner#1478]
**Describe the enhancement**

Currently, three Action types are supported: Docker, JavaScript and Composite (see https://docs.github.com/en/actions/creating-actions/about-custom-actions#types-of-actions). However, features `pre`, `pref-if`, `post` and `post-if` are only is supported in JavaScript Actions only (see https://docs.github.com/en/actions/creating-actions/metadata-syntax-for-github-actions#runs-for-javascript-actions). Therefore, users writing workflows/actions using some scripting language such as Python are forced to wrap the steps in JavaScript in order to register pre and post steps. See, for example, https://github.com/hdl/containers/tree/main/utils/registry-login:

```yml
name: 'Login to container registries and set a post step'

inputs:
  precmd:
    description: 'Pre command'
    required: false
    default: 'precmd.py'
  postcmd:
    description: 'Post command'
    required: false
    default: 'postcmd.py'

runs:
  using: 'node12'
  main: 'main.js'
  post: 'main.js'
```

```js
const { exec } = require('child_process');

function run(cmd) {
  exec(cmd, (error, stdout, stderr) => {
    if ( stdout.length != 0 ) { console.log(`${stdout}`); }
    if ( stderr.length != 0 ) { console.error(`${stderr}`); }
    if (error) {
      console.error(`exec error: ${error}`);
    }
    return error;
  });
}

// Are we in the 'post' step?
if ( process.env.STATE_FLAG != undefined ) {
  run(process.env.INPUT_POSTCMD);
  return;
}

// Otherwise, this is the main step
run(process.env.INPUT_PRECMD);
console.log(`::save-state name=FLAG::placeholder`);
```

The complexity might be simplified if Python or Bash or PowerShell were supported similarly to JavaScript:

```yml
name: 'Login to container registries and set a post step'

runs:
  using: 'python'
  main: 'precmd.py'
  post: 'postcmd.py'
```

**Additional information**

Alternatively, since regular steps support Python as a built-in shell already, the same capability might be achieved if Composite Actions supported fields `pre` and `post` as a complement to `steps`. For instance:

```yml
name: 'Login to container registries and set a post step'

inputs:
  precmd:
    description: 'Pre command'
    required: true
  postcmd:
    description: 'Pre command'
    required: true

runs:
  using: 'composite'
  pre:
    - shell: python
      run: ${{ inputs.precmd }}
  post:
    - shell: python
      run: ${{ inputs.postcmd }}
  steps:
    - ...
```

/cc @thboop, per https://github.com/actions/runner/issues/646#issuecomment-901336347

**Comment by webknjaz** [comment]
I've also faced this need. Subscribing.

**Comment by umarcor** [comment]
@webknjaz you might want to use `hdl/containers/utils/with-post-step`, or just copy the sources to you own repo: https://github.com/hdl/containers/tree/main/utils/with-post-step. For instance:

```yml
  - name: Release
    uses: ./utils/with-post-step
    with:
      main: |
        echo '${{ inputs.gcr_token }}' | docker login gcr.io -u _json_key --password-stdin
        echo '${{ inputs.gh_token }}' | docker login ghcr.io -u gha --password-stdin
        echo '${{ inputs.docker_pass }}' | docker login docker.io -u '${{ inputs.docker_user }}' --password-stdin
        dockerRelease ${{ inputs.architecture }} ${{ inputs.collection }} ${{ inputs.images }}
      post: for registry in gcr.io ghcr.io docker.io; do docker logout "$registry"; done
```

**Comment by webknjaz** [comment]
@umarcor I was considering something like this. Will it run at the end of the job execution or at the composite action exit?

**Comment by umarcor** [comment]
When I use it in a composite action, I call that action once only. Therefore, there is no difference between the end of the composite action or the end of the job. However, according to documentation, all post steps are executed at the end of the job, in reverse order.

**Comment by ethomson** [comment]
Supporting `pre` and `post` in composite actions seems like a usable addition.  We don't have this on our roadmap at the moment, but I'll add it to the list for future work.  Thanks for the suggestion!

**Comment by jessehouwing** [comment]
I have a use-case for this as well having a `runs-post: |` block would also work for me, that's basically the trick these custom post-step actions employ. But then you get to pick the kind of script host and add environment variables and such.

**Comment by jessehouwing** [comment]
> @webknjaz you might want to use `hdl/containers/utils/with-post-step`, or just copy the sources to you own repo: https://github.com/pyTooling/Actions/tree/main/with-post-step. For instance:
> 
> ```yaml
>   - name: Release
>     uses: ./utils/with-post-step
>     with:
>       main: |
>         echo '${{ inputs.gcr_token }}' | docker login gcr.io -u _json_key --password-stdin
>         echo '${{ inputs.gh_token }}' | docker login ghcr.io -u gha --password-stdin
>         echo '${{ inputs.docker_pass }}' | docker login docker.io -u '${{ inputs.docker_user }}' --password-stdin
>         dockerRelease ${{ inputs.architecture }} ${{ inputs.collection }} ${{ inputs.images }}
>       post: for registry in gcr.io ghcr.io docker.io; do docker logout "$registry"; done
> ```

Anyone know what he magic syntax would be to include such local action in the composite action repo?

```
    - name: Wait for user to terminate workflow
      uses: ./with-post-step
      with:
```

in the action YAML will be relative to therepo root, not relative to the composite's `action.yaml` file.

**Comment by umarcor** [comment]
Maybe `uses: ${{ github.action_path }}/with-post-step`? (I did not try it)

**Comment by jessehouwing** [comment]
Tried that. Not making actions happy:

```
Error: jessehouwing/debug-via-ssh/main/action.yaml (Line: 154, Col: 13):
Error: jessehouwing/debug-via-ssh/main/action.yaml (Line: 154, Col: 13): Unrecognized named-value: 'github'. Located at position 1 within expression: github.action_path
Error: jessehouwing/debug-via-ssh/main/action.yaml (Line: 154, Col: 13): Expected format {org}/{repo}[/path]@ref. Actual '${{ github.action_path }}/with-post-step'
Error: System.FormatException: Input string was not in a correct format.
```

For now fixed by passing in the full action path:

```
uses: jessehouwing/debug-via-ssh/with-post-step@main
```
```

**Comment by umarcor** [comment]
@jessehouwing see https://github.com/github/feedback/discussions/9049 (via https://github.com/hdl/containers/issues/48).

**Comment by Kurt-von-Laven** [comment]
This would make it a lot easier to create quick and easy actions that cache things for various common tools/languages. Excited to hear it's already on the roadmap!

**Comment by menasheh** [comment]
When is this planned? Seems like a fundamental feature to me

**Comment by harzallah** [comment]
+1

**Comment by noahsbwilliams** [comment]
Just writing in support - would have a lot of value in the context of "write secret --> $action --> remove secret" tasks

**Comment by fortunacio** [comment]
??

**Comment by fabasoad** [comment]
> Supporting `pre` and `post` in composite actions seems like a usable addition. We don't have this on our roadmap at the moment, but I'll add it to the list for future work. Thanks for the suggestion!

Hi @ethomson! Do you have any news for us? Had it been added to the roadmap (if yes, maybe you could share the ETA) or not yet? Thanks.

**Comment by scottgigante-immunai** [comment]
+1 for this being a useful feature!

**Comment by Himura2la** [comment]
It's so strange that this basic feature is not implemented. I don't want to use JS for writing basic thing easily done in shell. What do you even mean by `composite` if it's simply a `shell` action??
The restriction on post-action is super strange, I even see this step in my workflow, but I can't add any code in it
![image](https://github.com/actions/runner/assets/7802808/32e445ca-d170-44a5-99c8-8f808a8c715d)

**Comment by antoineco** [comment]
@Himura2la everybody here wants to see this happen as much as you do. Until this happens, you can use this (as already suggested) so that you don't have to write any Javascript yourself:

```yaml
    - id: bootstrap
      uses: pyTooling/Actions/with-post-step@v0.4.5
      with:
        main: |
          main shell commands
        post: |-
          post execution shell commands
```

**Comment by hstravis** [comment]
We've been doing this as the last step in our composite jobs:

```yaml
    - name: Cleanup after main action
      if: ${{ always() }}
      shell: bash
      run: |
        # clean up script goes here
```

**Comment by umarcor** [comment]
> It's the same as the action shared by @antoineco but doesn't require a public action - some companies have restrictions on public actions.

For completeness, they can copy the sources to their repo and use it either internally or locally. See https://github.com/pyTooling/Actions/tree/main/with-post-step. It's 10 lines of Apache licensed code.

**Comment by philomory** [comment]
The `with-post-step` action is a useful stopgap, but, besides not being as "clean" as a native solution, it's worth pointing out that there are actually situations where it doesn't actually work (and as far as I know, could not be made to work); specifically, if you want to schedule an _action_ (rather than a shell script) to run as post-step of your composite action, `with-post-step` won't actually help.

The place that this has come up most frequently for me is wanting to have a composite action that executes `actions/cache/restore` as part of the composite action, and then schedules `actions/cache/save` as a post-run step (that runs at the end of the calling workflow, not at the end of the composite action). Using `actions/cache/restore` and `actions/cache/save` individually gives you a lot more control than just using `actions/cache` (which is why they exist in the first place), but, using them as part of a composite action is pretty unwieldy, and I don't think it's actually possible to fix that through clever workarounds like `with-post-step`, it's something that would need native support.

**Comment by sandstrom** [comment]
This would solve a lot of headaches around cache logic and such. Especially since it's not possible to use post steps in Reusable Workflows either.

**Comment by zak905** [comment]
Hello, any updates about this ?

**Comment by junaruga** [comment]
We, the Ruby project is using the [gacts/run-and-post-run GitHub Action](https://github.com/gacts/run-and-post-run) as alternative at [the commit](https://github.com/ruby/ruby/commit/8e3cb69a83fea488d7fa279a872fba73326d2ed2#diff-4e89dc03d96d958e294074baa5d905cf68a3da2840e299519c06688f8d5aaca8R138).

**Comment by rrauenza** [comment]
Another workaround, add this to the top of your run:

```
- name: run container 
  run: |
     function cleanup() {                                                
          docker image rm "${{ steps.build-image.outputs.image }}" || true        
      }                                                                       
      trap cleanup EXIT   
      docker run "${{ steps.build-image.outputs.image }}"
```

**Comment by jaymecd** [comment]
`trap` works within same shell.

so this is almost equivalent (except for `run` fails):
```shell
      docker run "${{ steps.build-image.outputs.image }}"
      docker image rm "${{ steps.build-image.outputs.image }}" || true        
```

**Comment by RockLobster** [comment]
> Another workaround, add this to the top of your run if it is sh/bash:
> 
> ```
> - name: run container 
>   run: |
>       function cleanup() {                                                
>           docker image rm "${{ steps.build-image.outputs.image }}" || true        
>       }                                                                       
>       trap cleanup EXIT   
>       docker run "${{ steps.build-image.outputs.image }}"
> ```
> 
> edit: Can anyone explain the thumbs down? Am I mistaken that this works to achieve a post cleanup that always runs?

The post step is intented to run once the whole github job is done, not just after the current step/action

**Comment by garrytrinder** [comment]
+1 would love to see pre and post steps added to composite actions

**Comment by takeda** [comment]
This issue is now 4 years old, it looks like it would be trivial to add, many people miss it and it was acknowledged that it would be "an usable addition". Why there's such long holdout on this?

**Comment by dkfellows** [comment]
> Why there's such long holdout on this?

Needs more ~cowbell~ copilot.

**Comment by ericsciple** [comment]
Hey folks 👋

I've been reviewing this thread and want to clarify something that might help scope this request.

**The cache use case already works today.** When you use `actions/cache@v5` (or any action with `pre:`/`post:`) inside a composite action, the runner automatically collects and executes those nested pre/post steps. For example:

```yaml
# my-caching-composite/action.yml
runs:
  using: composite
  steps:
    - uses: actions/cache@v5
      with:
        path: ~/.npm
        key: ${{ runner.os }}-npm-${{ hashFiles('**/package-lock.json') }}
    - run: npm ci
```

The cache's `post:` step (save) runs automatically after the job completes. You don't need explicit `pre:`/`post:` blocks in your composite action for this pattern.

**Question for the thread:** Given this, would support for just `run` steps in `pre:`/`post:` blocks be sufficient for your use cases? For example:

```yaml
runs:
  using: composite
  pre:
    - run: echo "Login to registry"
    - run: docker login ...
  steps:
    - run: ./main-logic.sh
  post:
    - run: docker logout
    - run: echo "Cleanup complete"
```

Or are there scenarios where you specifically need `uses:` in `pre:`/`post:` that **cannot** be achieved by nesting the action in `steps:` (like the cache example above)?

I'm trying to understand if there's an incremental approach (`run`-only first) that would address most use cases, versus needing full `run` + `uses` support from the start.

**Comment by ljharb** [comment]
It would certainly address many of mine, but literally any place I'd put steps/runs is a place i'd eventually want to abstract to a shared component - Tennant's correspondence principle and all that.

**Comment by tianon** [comment]
My biggest use case is writing shared actions in pure shell script, and composite being the only way to really make that work across all platforms without writing Node.js (it feels ironic coming from me in particular, but I find Docker-based actions awful, and they're even worse on Windows).

So from that perspective, starting with `run:` would be fantastic, especially if it's done in a way that allows it to expand to include `uses:` in some future update (ie, the proposed syntax seems great to me). 👍

**Edit:** being totally honest, I only personally need `post:`, since anything in `pre:` should be fine to live in `steps:` anyhow

**Comment by ataylorme** [comment]
`run:` pre and post steps would support my use case of sending telemetry data at the end of a composite action in a post step

**Comment by DekusDenial** [comment]
`run` step in pre/post block would address most of my use cases where currently are implemented by separating steps into other actions which would have been better off inside just one with pre/post

**Comment by basilevs** [comment]
I need to upload artifacts in the post step, so `run ` is probably insufficient.

**Comment by stevehipwell** [comment]
Being able to only use `run:` in `pre:` & `post:` works for me. I'm not sure that supporting `uses:` would be a good idea, but it seems like another discussion as it'd be new functionality compared to `runs:` which provides the same functionality as a node action.

**Comment by dkfellows** [comment]
There are two types of use cases for post-actions:

1. **Always run**: for handling things where you want to run them even if a subsequent step fails.
2. **Conditional run**: for handling things where you only want to run them on success.

## Examples
The first case is useful for cleaning up some types of security token. The second case is useful for logging execution metrics.

**Comment by jsoref** [comment]
I'm pretty sure I'd want `timeout-minutes` and `if` for pre/post steps.

I regularly use upload-artifact and have historically used upload-sarif, but in general, I'm not sure if I'd need them for anything today. I can probably live without access to other actions, if I need to, I can unroll things myself or borrow the actions runtime token and then inline the code that github actions does (worst case using act or a relative).

**Comment by thisisthetechie** [comment]
I'll add my voice to this for my immediate use case. 

I use custom actions quite a lot, it helps to have a shared library of actions that can be run within multiple Workflows without the need to re-write the wheel, so to speak.

One specific case for this is when dealing with Terraform. I have a set of custom actions that cater to how our Terraform is configured for our project. One issue I have to deal with is if the TF apply fails for any reason, be it a timeout or the job is cancelled mid-apply. When this happens, the action is just abandoned, with no way of cleaning up natively.

Adding the abiliity to include `post:` actions within the composite action that is handled by GitHub Actions as a proper "post action" step (i.e. is not a part of the action step, but runs every time that step is exited for whatever reason) then I can easily include the required activity to clean up a failed run - in this case, I can delete the state lock ready for the next run against that environment.

This has a much broader use case, for instance I would be able to send notifications and pre-configure my credential steps in the actual action, rather than having to have repeated steps to handle these items within the main workflow files. By using `pre` and `post` action declarations I can tidy up a lot of existing workflows to just contain the activities they perform knowing that the environments in which they are performing them are handled within the action.

I know there are other workarounds and ways of doing this, but these typically require multiple `always()`, `!success()`, `cancelled()` statements per step/job which is just messy. I'd much prefer to be able to configure my custom action to handle things before and after the main step code, using the actual GitHub Actions stages for these activities so that they are handled outside of the action itself just like the node actions are (without having to write a load of node/typescript nonsense for things that aren't in any way required).

And, to re-iterate, will run even if the job times out or is cancelled.

**Comment by joffrey-bion** [comment]
@ericsciple Thanks for trying to find some path to resolution at last!

> Question for the thread: Given this, would support for just run steps in pre:/post: blocks be sufficient for your use cases?

That wouldn't help me much. In post, I want to:

* `upload-artifacts`
* set some job summary / PR checks (e.g. JUnit test results)

I almost never need plain `run`.

**Comment by joffrey-bion** [comment]
> I'm trying to understand if there's an incremental approach (run-only first) that would address most use cases, versus needing full run + uses support from the start.

@ericsciple While I welcome the willingness to move forward _somehow_, I wonder though. Why is it so hard to add a simple group of `post` steps in composite actions? As in:
```yaml
name: 'My composite action with post steps'
runs:
  using: composite
  steps:
    - uses: ...
  post-steps:
    - uses: ...
```
And then, when a user calls this composite action from a workflow, you run all the regular `steps` like you do now, and then add a mirror post composite action that runs all the `post-steps` at the end of the workflow (in the proper reverse-ordered place within other post steps). Is it that much more complicated than it looks?

I guess the main workaround for us mere mortals (I imagine) is to create 2 composite actions instead of one, and use them manually at the beginning and end of each workflow, which is not perfect, because the `post-steps` are potentially not in the right place with respect to other post steps from the workflow. 
But this doesn't work for people who want to distribute proper actions.

**Comment by philomory** [comment]
> Hey folks 👋
> 
> I've been reviewing this thread and want to clarify something that might help scope this request.
> 
> **The cache use case already works today.** When you use `actions/cache@v5` (or any action with `pre:`/`post:`) inside a composite action, the runner automatically collects and executes those nested pre/post steps. For example:
> 
> # my-caching-composite/action.yml
> runs:
>   using: composite
>   steps:
>     - uses: actions/cache@v5
>       with:
>         path: ~/.npm
>         key: ${{ runner.os }}-npm-${{ hashFiles('**/package-lock.json') }}
>     - run: npm ci
> 
> The cache's `post:` step (save) runs automatically after the job completes. You don't need explicit `pre:`/`post:` blocks in your composite action for this pattern.

The cache `post` step runs automatically, with a lot of assumptions, after the job completes successfully. There several configuration tweaks you can make to cache saving that can _only_ be made if you explicitly invoke `actions/cache/restore` and `actions/cache/save` individually, because the `actions/cache` wrapper doesn't expose all of that functionality. For example, sometimes you might want to cache dependencies even if the actual build job failed, but the `actions/cache` wrapper doesn't work that way: the README from `actions/cache` directly suggests directly invoking `actions/cache/restore` and `actions/cache/save` in that case, instead. So saying "the cache use case already works today" sort of misses the point of all of the posts talking about the cache actions further up the thread.
